src/pg_copy.c

```diff
@@ -0,0 +1,600 @@
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+		}
+
+		while (true)
+		{
+			ShardInterval *shardInterval = NULL;
+			ShardId shardId = 0;
+			bool found = false;
+
+			MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+			nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+			MemoryContextSwitchTo(oldContext);
+		
+			if (!nextRowFound)
+			{
+				MemoryContextReset(tupleContext);
+				break;
+			}
+			
+			/* write the row to the shard */
+			if (columnNulls[partitionColumn->varattno]) 
+			{ 
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			} 
+			partitionColumnValue = columnValues[partitionColumn->varattno];
+			hashedValue = DatumGetInt32(FunctionCall1(hashFunction, partitionColumnValue));
+			shardHashCode = (int)((uint32)(hashedValue - INT32_MIN) / hashTokenIncrement);
+			prunedList = shardListCache[shardHashCode];
+			
+			if (prunedList == NULL)
+			{
+				rightConst->constvalue = partitionColumnValue;
+				prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+				shardListCache[shardHashCode] = prunedList;
+			}
+
+			shardInterval = (ShardInterval *) lfirst(list_head(prunedList));
+			shardId = shardInterval->id;
+				
+			shardConnections = (ShardConnections*)hash_search(shardToConn, &shardInterval->id, HASH_ENTER, &found);
+			if (!found) 
+			{ 
+				InitializeShardConnections(copyStatement, shardConnections, shardId, transactionManager);
+			}
+			lineBuf = CopyGetLineBuf(copyState);
+			lineBuf->data[lineBuf->len++] = '\n'; 
+			/* There was already new line in the buffer, but it was truncated: 
+			 * no need to check available space */
+			
+			/* Replicate row to all shard placements */
+			for (i = 0; i < shardConnections->replicaCount; i++) 
+			{ 
+				if (PQputCopyData(shardConnections->conn[i], lineBuf->data, lineBuf->len) <= 0)
+				{
+					elog(ERROR, "Copy failed for shard %ld", (long)shardId);
+				}
+			}
+			MemoryContextReset(tupleContext);
+		}
+		/* Perform two phase commit in replicas */
+		shardConnectionsList = HTABToList(shardToConn);
+		failedShard = DoForAllShards(shardConnectionsList, PgCopyPrepareTransaction, NULL);
+	}
+	PG_CATCH(); /* do recovery */
+	{
+		EndCopyFrom(copyState);
+		heap_close(rel, AccessShareLock);
+		/* Rollback transactions */
+		shardConnectionsList = HTABToList(shardToConn);
+		DoForAllShards(shardConnectionsList, PgCopyAbortTransaction, NULL);
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	EndCopyFrom(copyState);
+	heap_close(rel, AccessShareLock);
+
+	/* Complete two phase commit */
+	if (failedShard != INVALID_SHARD_ID) 
+	{ 
+		DoForAllShards(shardConnectionsList, PgCopyAbortTransaction, NULL);
+		elog(ERROR, "COPY failed for shard %ld", (long)failedShard);
```

Comment:
This code path appears to be identical to adding:

    if (failedShard != INVALID_SHARD_ID) 
    {
        elog(ERROR, "COPY failed for shard %ld", (long)failedShard);
    }

before line 574, since the clean-up/abort code will be executed in the PG_CATCH block.